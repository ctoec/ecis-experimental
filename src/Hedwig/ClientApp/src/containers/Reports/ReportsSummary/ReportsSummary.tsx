import React, { useContext } from 'react';
import { Link } from 'react-router-dom';
import { Table, TableProps } from '../../../components/Table/Table';
import dateFormatter from '../../../utils/dateFormatter';
import UserContext from '../../../contexts/User/UserContext';
import useApi from '../../../hooks/useApi';
import {
	ApiOrganizationsOrgIdReportsGetRequest,
	CdcReport as Report,
	ApiOrganizationsIdGetRequest,
} from '../../../generated';
import { getIdForUser, reportingPeriodFormatter } from '../../../utils/models';
import { DeepNonUndefineable } from '../../../utils/types';
import CommonContainer from '../../CommonContainer';

export default function ReportsSummary() {
	const { user } = useContext(UserContext);
	const orgParams: ApiOrganizationsIdGetRequest = {
		id: getIdForUser(user, 'org'),
	};
	const [orgLoading, orgError, organization] = useApi(api => api.apiOrganizationsIdGet(orgParams), [
		user,
	]);
	const reportParams: ApiOrganizationsOrgIdReportsGetRequest = {
		orgId: getIdForUser(user, 'org'),
	};
	const [loading, error, reports] = useApi(
		api => api.apiOrganizationsOrgIdReportsGet(reportParams),
		[user]
	);

	if (loading || error || !reports || orgLoading || orgError || !organization) {
		return <div className="Reports"></div>;
	}

	const pendingReports = reports.filter<DeepNonUndefineable<Report>>(
		(r => !r.submittedAt) as (_: DeepNonUndefineable<Report>) => _ is DeepNonUndefineable<Report>
	);

	const defaultTableProps: TableProps<DeepNonUndefineable<Report>> = {
		id: 'reports-table',
		data: reports,
		rowKey: row => row.id,
		columns: [
			{
				name: 'Period',
				cell: ({ row }) => (
					<th scope="row">
						<Link to={`/reports/${row.id}`} className="usa-link">
							{reportingPeriodFormatter(row.reportingPeriod)}
						</Link>
					</th>
				),
				sort: row => row.reportingPeriod.period.getTime() || 0,
				width: '21%',
			},
			{
				name: 'Type',
				cell: ({ row }) => <td>{row.type}</td>,
				sort: row => row.type,
				width: '16%',
			},
			{
				name: 'Program/Site',
				cell: _ => <td>{organization.name}</td>,
				sort: _ => organization.name || '',
				width: '38%',
			},
		],
		defaultSortColumn: 0,
		defaultSortOrder: 'descending',
	};

	const pendingTableProps: TableProps<DeepNonUndefineable<Report>> = {
		...defaultTableProps,
		id: 'pending-reports-table',
		data: pendingReports,
		columns: [
			...defaultTableProps.columns,
			{
				name: 'Due date',
				cell: ({ row }) => (
					<td className="oec-table__cell--tabular-nums oec-table__cell--gray">
						{dateFormatter(row.reportingPeriod.dueAt)}
					</td>
				),
				sort: row => row.reportingPeriod.dueAt.getTime() || 0,
				width: '24%',
			},
		],
		defaultSortColumn: 3,
		defaultSortOrder: 'ascending',
	};

	const submittedTableProps: TableProps<DeepNonUndefineable<Report>> = {
		...defaultTableProps,
		id: 'submitted-reports-table',
		data: reports.filter<DeepNonUndefineable<Report>>(
			(r => !!r.submittedAt) as (_: DeepNonUndefineable<Report>) => _ is DeepNonUndefineable<Report>
		),
		columns: [
			...defaultTableProps.columns,
			{
				name: 'Date submitted',
				cell: ({ row }) => (
					<td className="oec-table__cell--tabular-nums">{dateFormatter(row.submittedAt)}</td>
				),
				sort: row => (row.submittedAt && row.submittedAt.getTime()) || 0,
				width: '24%',
			},
		],
	};

	return (
		<CommonContainer>
			<div className="grid-container">
				<h1>Reports</h1>
				<section>
					<h2 className="margin-top-4">Pending reports</h2>
					{pendingReports.length > 0 ? (
						<Table {...pendingTableProps} fullWidth />
					) : (
						<p>
							<em>
								No reports pending. Reports become available after the end of the reporting period.
							</em>
						</p>
					)}
				</section>
				<section>
					<h2 className="margin-top-6">Submitted reports</h2>
					<Table {...submittedTableProps} fullWidth />
				</section>
			</div>
		</CommonContainer>
	);
}
