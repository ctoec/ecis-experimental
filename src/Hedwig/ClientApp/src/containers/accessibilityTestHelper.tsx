import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';


expect.extend(toHaveNoViolations);

// This is a *very* minimal accessibility check
// From the README for the jest-axe library: "The GDS Accessibility team found that only ~30% of issues are found by automated testing."
export function accessibilityTestHelper(testComponent: React.ReactElement) {
	it('passes accessibility checks', async () => {
		const { container } = render(testComponent);
		const results = await axe(container);
		expect(results).toHaveNoViolations();
		cleanup();
	});
}
