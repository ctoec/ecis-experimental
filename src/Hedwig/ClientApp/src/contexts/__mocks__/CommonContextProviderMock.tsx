import React from 'react';
import { Router } from 'react-router-dom';
import UserContext from '../User/UserContext';
import ReportingPeriodContext from '../ReportingPeriod/ReportingPeriodContext';
import { User, Region, ReportingPeriod, FundingSource } from '../../generated';
import emptyGuid from '../../utils/emptyGuid';
import { createMemoryHistory, History } from 'history';

export const defaultUser: User = {
	id: 1,
	wingedKeysId: emptyGuid(),
	firstName: 'Minerva',
	lastName: 'McGonagall',
	orgPermissions: [
		{
			organizationId: 1,
			organization: {
				id: 1,
				name: "Children's Adventure Center",
				sites: [
					{
						id: 1,
						name: "Children's Adventure Center",
						organizationId: 1,
						region: Region.East,
						titleI: false,
					},
				],
			},
			id: 1,
			userId: 1,
		},
	],
	sitePermissions: [],
};

export const defaultCdcReportingPeriods: ReportingPeriod[] = [
	{
		id: 1,
		type: FundingSource.CDC,
		period: new Date('2019-03-01'),
		periodStart: new Date('2019-03-01'),
		periodEnd: new Date('2019-03-31'),
		dueAt: new Date('2019-04-15'),
	},
	{
		id: 2,
		type: FundingSource.CDC,
		period: new Date('2019-04-01'),
		periodStart: new Date('2019-04-01'),
		periodEnd: new Date('2019-04-30'),
		dueAt: new Date('2019-05-15'),
	},
	{
		id: 3,
		type: FundingSource.CDC,
		period: new Date('2019-05-01'),
		periodStart: new Date('2019-05-01'),
		periodEnd: new Date('2019-05-31'),
		dueAt: new Date('2019-06-15'),
	},
];

type CommonContextProviderMockProps = {
	user?: User;
	cdcReportingPeriods?: ReportingPeriod[];
	history?: History<any>;
};

const CommonContextProviderMock: React.FC<CommonContextProviderMockProps> = ({
	children,
	user = defaultUser,
	cdcReportingPeriods = defaultCdcReportingPeriods,
	history = createMemoryHistory(),
}) => {
	return (
		<UserContext.Provider value={{ user }}>
			<ReportingPeriodContext.Provider value={{ cdcReportingPeriods }}>
				<Router history={history}>{children}</Router>
			</ReportingPeriodContext.Provider>
		</UserContext.Provider>
	);
};

export default CommonContextProviderMock;
