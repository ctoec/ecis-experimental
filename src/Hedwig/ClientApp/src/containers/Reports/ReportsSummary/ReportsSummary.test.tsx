// Variables used in jest mockes -- must start with `mock`
import {
	mockDefaultReport,
	mockReport as _mockReport,
	mockCompleteEnrollment,
	mockEnrollmentWithFoster,
	mockOrganization,
} from '../../../tests/data';
import mockUseApi, { mockApiOrganizationsIdGet } from '../../../hooks/__mocks__/useApi';

const submittedReport = { ...mockDefaultReport, submittedAt: new Date('2019-09-14') };
const defaultReports = [mockDefaultReport, submittedReport];

let mockReports = defaultReports;

// Jest mocks must occur before later imports
jest.mock('../../../hooks/useApi', () =>
	mockUseApi({
		apiOrganizationsIdGet: mockApiOrganizationsIdGet(mockOrganization),
		apiOrganizationsOrgIdReportsGet: (_: any) => [false, null, mockReports],
	})
);

import React from 'react';
import { render } from '@testing-library/react';
import ReportsSummary from './ReportsSummary';
import CommonContextProviderMock from '../../../contexts/__mocks__/TestProvider';
import { accessibilityTestHelper } from '../../accessibilityTestHelper';

afterAll(() => {
	jest.resetModules();
});

describe('ReportsSummary', () => {
	it('matches snapshot', () => {
		const { asFragment } = render(
			<CommonContextProviderMock>
				<ReportsSummary />
			</CommonContextProviderMock>
		);
		expect(asFragment()).toMatchSnapshot();
	});

	describe('when there are no pending reports', () => {
		beforeEach(() => {
			mockReports = [submittedReport];
		});

		it('shows explanatory text', () => {
			const { getAllByText } = render(
				<CommonContextProviderMock>
					<ReportsSummary />
				</CommonContextProviderMock>
			);
			expect(getAllByText(/No reports pending/)).toHaveLength(1);
		});
	});

	accessibilityTestHelper(
		<CommonContextProviderMock>
			<ReportsSummary />
		</CommonContextProviderMock>
	);
});
