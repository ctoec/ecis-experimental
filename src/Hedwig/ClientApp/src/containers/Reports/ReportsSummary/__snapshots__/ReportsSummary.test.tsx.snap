// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Roster matches snapshot 1`] = `"<div class=\\"grid-container\\"></div><div class=\\"grid-container\\"><h1>Reports</h1><section><h2 class=\\"margin-top-4\\">Pending reports</h2><table id=\\"pending-reports-table\\" class=\\"oec-table oec-table--full-width\\"><thead><tr><th scope=\\"col\\" class=\\"oec-table__column-header oec-sortable\\" role=\\"columnheader\\" aria-sort=\\"descending\\"><button class=\\"oec-table__column-title usa-button--unstyled width-full\\" aria-label=\\"Sort table by Period in ascending order\\">Period<div class=\\"oec-table__sort-controls\\"><span class=\\"oec-table__sort-controls__asc\\"></span><span class=\\"oec-table__sort-controls__desc active\\"></span></div></button></th><th scope=\\"col\\" class=\\"oec-table__column-header oec-sortable\\" role=\\"columnheader\\" aria-sort=\\"descending\\"><button class=\\"oec-table__column-title usa-button--unstyled width-full\\" aria-label=\\"Sort table by Type in ascending order\\">Type<div class=\\"oec-table__sort-controls\\"><span class=\\"oec-table__sort-controls__asc\\"></span><span class=\\"oec-table__sort-controls__desc\\"></span></div></button></th><th scope=\\"col\\" class=\\"oec-table__column-header oec-sortable\\" role=\\"columnheader\\" aria-sort=\\"descending\\"><button class=\\"oec-table__column-title usa-button--unstyled width-full\\" aria-label=\\"Sort table by Program/Site in ascending order\\">Program/Site<div class=\\"oec-table__sort-controls\\"><span class=\\"oec-table__sort-controls__asc\\"></span><span class=\\"oec-table__sort-controls__desc\\"></span></div></button></th><th scope=\\"col\\" class=\\"oec-table__column-header oec-sortable\\" role=\\"columnheader\\" aria-sort=\\"descending\\"><button class=\\"oec-table__column-title usa-button--unstyled width-full\\" aria-label=\\"Sort table by Due date in ascending order\\">Due date<div class=\\"oec-table__sort-controls\\"><span class=\\"oec-table__sort-controls__asc\\"></span><span class=\\"oec-table__sort-controls__desc\\"></span></div></button></th></tr></thead><tbody><tr><th scope=\\"row\\"><a class=\\"usa-link\\" href=\\"/reports/2\\">September 2019</a></th><td>CDC</td><td>Children's Adventure Center</td><td class=\\"oec-table__cell--tabular-nums oec-table__cell--gray\\">10/15/2019</td></tr></tbody></table></section><section><h2 class=\\"margin-top-6\\">Submitted reports</h2><table id=\\"submitted-reports-table\\" class=\\"oec-table oec-table--full-width\\"><thead><tr><th scope=\\"col\\" class=\\"oec-table__column-header oec-sortable\\" role=\\"columnheader\\" aria-sort=\\"descending\\"><button class=\\"oec-table__column-title usa-button--unstyled width-full\\" aria-label=\\"Sort table by Period in ascending order\\">Period<div class=\\"oec-table__sort-controls\\"><span class=\\"oec-table__sort-controls__asc\\"></span><span class=\\"oec-table__sort-controls__desc active\\"></span></div></button></th><th scope=\\"col\\" class=\\"oec-table__column-header oec-sortable\\" role=\\"columnheader\\" aria-sort=\\"descending\\"><button class=\\"oec-table__column-title usa-button--unstyled width-full\\" aria-label=\\"Sort table by Type in ascending order\\">Type<div class=\\"oec-table__sort-controls\\"><span class=\\"oec-table__sort-controls__asc\\"></span><span class=\\"oec-table__sort-controls__desc\\"></span></div></button></th><th scope=\\"col\\" class=\\"oec-table__column-header oec-sortable\\" role=\\"columnheader\\" aria-sort=\\"descending\\"><button class=\\"oec-table__column-title usa-button--unstyled width-full\\" aria-label=\\"Sort table by Program/Site in ascending order\\">Program/Site<div class=\\"oec-table__sort-controls\\"><span class=\\"oec-table__sort-controls__asc\\"></span><span class=\\"oec-table__sort-controls__desc\\"></span></div></button></th><th scope=\\"col\\" class=\\"oec-table__column-header oec-sortable\\" role=\\"columnheader\\" aria-sort=\\"descending\\"><button class=\\"oec-table__column-title usa-button--unstyled width-full\\" aria-label=\\"Sort table by Date submitted in ascending order\\">Date submitted<div class=\\"oec-table__sort-controls\\"><span class=\\"oec-table__sort-controls__asc\\"></span><span class=\\"oec-table__sort-controls__desc\\"></span></div></button></th></tr></thead><tbody><tr><th scope=\\"row\\"><a class=\\"usa-link\\" href=\\"/reports/1\\">August 2019</a></th><td>CDC</td><td>Children's Adventure Center</td><td class=\\"oec-table__cell--tabular-nums\\">09/09/2019</td></tr></tbody></table></section></div>"`;
